scripts/opportunity-provider/index.ts

```diff
@@ -0,0 +1,223 @@
+        {
+          token: opportunity.buyToken,
+          amount: opportunity.buyAmount,
+        },
+      ],
+      owner: account.address,
+    },
+  };
+
+  return account.signTypedData({
+    domain: {
+      name: "Permit2",
+      verifyingContract: config.permit2 as Address,
+      chainId: config.chainNetworkId,
+    },
+    types,
+    primaryType: "PermitBatchWitnessTransferFrom",
+    message,
+  });
+}
+
+function getCallData(
+  account: PrivateKeyAccount,
+  opportunity: Opportunity,
+  nonce: number,
+  deadline: number,
+  signature: `0x${string}`
+) {
+  return encodeFunctionData({
+    abi,
+    functionName: "execute",
+    args: [
+      {
+        permit: {
+          permitted: [
+            {
+              token: opportunity.sellToken as Address,
+              amount: BigInt(opportunity.sellAmount),
+            },
+          ],
+          nonce: BigInt(nonce),
+          deadline: BigInt(deadline),
+        },
+        witness: {
+          buyTokens: [
+            {
+              token: opportunity.buyToken as Address,
+              amount: BigInt(opportunity.buyAmount),
+            },
+          ],
+          owner: account.address,
+        },
+      },
+      signature,
+    ],
+  });
+}
+
+async function submitOpportunity(account: PrivateKeyAccount) {
+  const config = readConfig();
+  const opportunity = readOpportunity();
+
+  const nonce = Math.floor(Math.random() * 2 ** 50);
+  const deadline = Math.floor(Date.now() / 1000) + 60 * 60 * 24;
+  const signature = await signOpportunity(
+    account,
+    config,
+    opportunity,
+    nonce,
+    deadline
+  );
+  const permissionKey = encodeAbiParameters(
+    [
+      { type: "address", name: "admin" },
+      { type: "bytes", name: "signature" },
+    ],
+    [account.address, signature]
+  );
+
+  const params = {
+    version: "v1",
+    permission_key: permissionKey,
+    chain_id: config.chainId,
+    target_contract: config.opportunityProvider,
+    target_calldata: getCallData(
+      account,
+      opportunity,
+      nonce,
+      deadline,
+      signature
+    ),
+    target_call_value: "0",
+    sell_tokens: [
+      {
+        token: opportunity.buyToken,
+        amount: opportunity.buyAmount,
+      },
+    ],
+    buy_tokens: [
+      {
+        token: opportunity.sellToken,
+        amount: opportunity.sellAmount,
+      },
+    ],
+  };
+
+  console.log("Submitting opportunity...");
+  const response = await fetch(`${config.serverUrl}/v1/opportunities`, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify(params),
+  });
+  console.log("Opportunity submitted", response.status);
+  const data = await response.json();
+  console.log(data);
+}
+
+const argv = yargs(hideBin(process.argv))
+  .option("private-key", {
+    description:
+      "Private key to provide opportunity in hex format with 0x prefix. e.g: 0xdeadbeef...",
+    type: "string",
+    demandOption: true,
+  })
```

Comment:
maybe add options for config and opportunity path with default values set to `config.json` and `opportunity.json`. I can imagine having multiple pre-built opportunities for testing different scenarios and I'd rather not change the same file over and over again.

Reply:
could also add multiple opportunities to `opportunity.json`. i guess it depends in part on whether we want to create the same set of opportunities again and again. if so, then i'd say just add to a single json.